### PR TITLE
TCX support

### PR DESCRIPTION
**Pull request by evanbarter**
Addresses #35.

Functionally the geometry stuff works well AFAIK. I've plotted it with Leaflet and compared against Strava, Garmin Connect maps.

I've included a bunch of sports performance-related attributes as properties, as they _may_ be useful to me, but ultimately I can pull those out if you're only looking to support mapping-related properties.

### CODE REVIEW

File: lib/tcx.js [hunk 1]
```diff
@@ -0,0 +1,207 @@
...
+              );
+            }
+          }
+        });
+    }
+  }
+  if (track.length === 0) return;
+  const properties = Object.assign(
+    getProperties(node)
+  );
```
Comment (tmcw) [inline lib/tcx.js:174]: This line doesn't do anything, as far as I can tell - we're not merging properties or making a clone of the object. Most likely the `Object.assign` can be removed and we're assigning `properties` to the return value of `getProperties`

File: lib/tcx.js [hunk 2]
```diff
@@ -0,0 +1,207 @@
...
+  const extensions = node.getElementsByTagNameNS(
+    "http://www.garmin.com/xmlschemas/ActivityExtension/v2",
+    "*"
+  );
+  attributeNames.forEach((attr) => {
+    const raw = get1(node, attr[0]);
+    if (raw !== null) {
+      prop[attr[1]] = parseFloat(nodeVal(raw));
+    } else if (extensions !== null) {
+      for (let i = 0; i < extensions.length; i++) {
```
Comment (tmcw) [inline lib/tcx.js:33]: This is looking a bit problematic: we're looping through `attributeNames`, then a nested loop through each element of `extensions` if the attribute is not found. In the worst case, where there are a lot of extensions, it's a bunch of tight loops of wasted work.

I'm not sure which case this is trying to handle - that `get1` might fail to get an element but getElementsByTagNameNS will succeed. Possibly we should _always_ use one or the other, or separate these loops?

File: lib/tcx.js [hunk 3]
```diff
@@ -0,0 +1,207 @@
...
+  const extensions = x.getElementsByTagNameNS(
+    "http://www.garmin.com/xmlschemas/ActivityExtension/v2",
+    "*"
+  );
+  if (extensions !== null) {
+    for (let i = 0; i < extensions.length; i++) {
+      const extension = extensions[i];
+      attributeNames
+        .map((r) => r[0])
+        .filter((n) => n === extension.localName)
```
Comment (tmcw) [inline lib/tcx.js:83]: This seems like what is desired is not to get all namespaced tags and filter them by attributeNames, but rather to call `getElementByTagNameNS` with each attributeName and the namespace.

File: lib/tcx.js [hunk 4]
```diff
@@ -0,0 +1,207 @@
...
+        .forEach((name) => {
+          const raw = nodeVal(extension);
+          if (raw !== null) {
+            const v = parseFloat(raw);
+            if (!isNaN(v)) {
```
Comment (tmcw) [inline lib/tcx.js:88]: can skip a step by calling `parseFloat(nodeVal(extension))` - null will parse to NaN.

File: lib/tcx.js [hunk 5]
```diff
@@ -0,0 +1,207 @@
...
+              result[name] = v;
+            }
+          }
+        });
+    }
+  }
+  return result;
+}
+
+function getPoints(node, pointname) {
```
Comment (tmcw) [inline lib/tcx.js:98]: `pointname` is, afaict, always `Trackpoint`, so I don't think it needs to be an argument.

File: lib/tcx.js [hunk 6]
```diff
@@ -0,0 +1,207 @@
...
+  if (l < 2) return {}; // Invalid line in GeoJSON
+  for (let i = 0; i < l; i++) {
+    const c = coordPair(pts[i]);
+    if (c === null) continue;
+    line.push(c.coordinates);
+    if (c.time) times.push(c.time);
+    attributeNames
+      .map((r) => r[0])
+      .forEach((name) => {
+        if (c[name] || extendedValues[name]) {
```
Comment (tmcw) [inline lib/tcx.js:113]: Looping over all potential attributes here seems wasteful (esp when done with .map, which will create a temporary array per iteration) - would be better to structure the coordinate output in a way that you could loop over the coordinate's extra properties.

**Comment by tmcw** [comment]
Sorry about the rather extensive comments - realizing that that attributeNames idea came into the GPX parser in a PR last year which I wasn't adequately cautious about 😉  - I can replace that contention in the GPX parser if it helps set an example for how the TCX parser may work.

**Comment by evanbarter** [comment]
No problem! I'll wait for your refactor and clean this up.

On review I am questioning whether the extensions are necessary at all for TCX. The GPX extensions make sense in the context of GeoJSON (e.g. line style), but the TCX ActivityPoint extension I am fetching not so much since they're basically just sports performance stats.

**Comment by tmcw** [comment]
Okay! Fixed up the GPX stuff, the PR for reference was at https://github.com/tmcw/togeojson/pull/38 - it's not a _ton_ of change but is very detail-oriented, trying to simplify things and make sure there are no performance pitfalls. If you can, that change is basically the direction (it saves from re-iterating over attributeValues, saves some lines of code, etc). If you're busy, I can probably do a PR into this one refactoring it with the same style.

**Comment by tmcw** [comment]
Merged with #39 - thanks! Rolling a release shortly.

**Pull request closed without merging**